### PR TITLE
Add affinity and anti-affinity support

### LINKED ISSUE #942: podAntiAffinity
**Tell us about your request**

Please implement `podAntiAffinity`.

I don't see anything about this on the roadmap. Is this planned?

(In fact, I see almost nothing on the [roadmap](https://github.com/aws/karpenter/projects/3) currently. Issue?)

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**

I want to deploy [k8ssandra](https://k8ssandra.io/) but cannot because it is using pod anti-affinity to prevent multiple pods from being scheduled on the same worker

> Cass Operator uses pod anti-affinity by default to ensure that Cassandra pods are isolated from one another.

and Karpenter refuses to schedule.

I suspect there are a lot of high-availability operators that will currently _prefer_ or _require_ podAntiAffinity.

Prometheus: https://github.com/prometheus-operator/kube-prometheus/issues/1537

**Are you currently working around this issue?**

`allowMultipleNodesPerWorker: true`

From: https://docs.k8ssandra.io/tasks/scale/

**Additional context**

What should these operators be using instead?

I see in the FAQ that this is not a priority for Karpenter but I think podAntiAffinity is fairly common among some popular operators, and it is not clear what they should be doing instead.

Some discussion here: https://github.com/aws/karpenter/pull/695#discussion_r715817587

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by tzneal**
**1. Issue, if available:**

Fixes #942 and #985 

**2. Description of changes:**

Adds support for pod affinity and anti-affinity

**3. How was this change tested?**

Unit testing and on my local EKS cluster.

**4. Does this change impact docs?**
- [X] Yes, PR includes docs updates
- [ ] Yes, issue opened: *link to issue*
- [ ] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/controllers/provisioning/scheduling/topology.go
Comment (tzneal) [inline pkg/controllers/provisioning/scheduling/topology.go:397]: This branch added the runtime.Must() to the result of NewRequirement, previously we just ignored the error assuming it couldn't happen.  It did, and we ended up just ignoring all match expressions.

File: pkg/controllers/provisioning/scheduling/topology.go
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/topology.go:404]: Isn't this the default behavior of what you were doing before?

Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/topology.go:404]: Nope, the strings are different:

```go
package metav1
const (
	LabelSelectorOpIn           LabelSelectorOperator = "In"
	LabelSelectorOpNotIn        LabelSelectorOperator = "NotIn"
	LabelSelectorOpExists       LabelSelectorOperator = "Exists"
	LabelSelectorOpDoesNotExist LabelSelectorOperator = "DoesNotExist"
)
```
```go
package selection
const (
	DoesNotExist Operator = "!"
	Equals       Operator = "="
	DoubleEquals Operator = "=="
	In           Operator = "in"
	NotEquals    Operator = "!="
	NotIn        Operator = "notin"
	Exists       Operator = "exists"
	GreaterThan  Operator = "gt"
	LessThan     Operator = "lt"
)
```

File: pkg/apis/provisioning/v1alpha5/requirements.go
Comment (ellistarn) [inline pkg/apis/provisioning/v1alpha5/requirements.go:115]: I think we should refactor requirements in general. Perhaps worth punting on this until then. Happy to leave this in, but it bloats an already massive PR.

File: pkg/controllers/provisioning/scheduling/topologynodefilter.go
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/topologynodefilter.go:32]:
```suggestion
func NewTopologyNodeFilter(p *v1.Pod) TopologyNodeFilter {
```

File: pkg/controllers/provisioning/scheduling/topologynodefilter.go
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/topologynodefilter.go:74]: Thoughts on parity with the Requirements interface?
```suggestion
func (t TopologyNodeFilter) Compatible(requirements v1alpha5.Requirements) bool {
```

File: pkg/controllers/provisioning/scheduling/topologynodefilter.go
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/topologynodefilter.go:85]: Optional, you could:
```
  return len(t) == 0
```

File: pkg/controllers/provisioning/scheduling/topologynodefilter.go
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/topologynodefilter.go:59]: Why not just use the Compatible interface?

File: pkg/controllers/provisioning/scheduling/topologynodefilter.go
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/topologynodefilter.go:51]: This is used in one place. Thoughts on removing this and just using the Compatible interface in that case? 
```
nodeFilter.Compatible(v1alpha5.NewLabelRequirements(node.Labels))
```

File: pkg/controllers/provisioning/scheduling/topologynodefilter.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/topologynodefilter.go:32]: `New.*` is typically reserved for methods that return a pointer interface while `Make.*` is for methods that return value types.  I'm good either way.

File: pkg/controllers/provisioning/scheduling/topologynodefilter.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/topologynodefilter.go:51]: The concept of a filter being "Compatible" with a node is less intuitive to me than the concept of a filter "Matching" a node.

File: pkg/controllers/provisioning/scheduling/topologynodefilter.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/topologynodefilter.go:59]: I don't think it makes as much sense to ask if a filter is 'Compatible' with something compared to asking if a filter 'Matches' something.

File: pkg/apis/provisioning/v1alpha5/requirements.go
Comment (tzneal) [reply pkg/apis/provisioning/v1alpha5/requirements.go:115]: It greatly affects performance, I'd like to leave it in this one for now, or possibly just cherry-pick it to main to reduce this PRs size.  Need to get the benchmark CI in as it's very easy to make a change and send scheduling performance through the floor.

File: pkg/controllers/provisioning/scheduling/topologynodefilter.go
Comment (ellistarn) [reply pkg/controllers/provisioning/scheduling/topologynodefilter.go:32]: I'm not familiar with this. Is this from the Go compiler?

File: pkg/controllers/provisioning/scheduling/topologynodefilter.go
Comment (ellistarn) [reply pkg/controllers/provisioning/scheduling/topologynodefilter.go:59]: The intent of this comment is to reduce the number of concepts that a reader needs to be aware of. Naming of those concepts can happen later. It's not clear to me that we have multiple concepts we're dealing with.

File: pkg/controllers/provisioning/scheduling/topologynodefilter.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/topologynodefilter.go:32]: That's one place, it's not 100% followed consistently but it seems odd to have a New.*() method that doesn't return a pointer which is why I went with Make()

Go compiler
```
src/cmd/compile/internal/syntax/pos.go:24:func MakePos(base *PosBase, line, col uint) Pos { return Pos{base, sat32(line), sat32(col)} }
src/cmd/internal/src/pos.go:381:func makeLicoRaw(line, col uint) lico {
src/reflect/value.go:2912:func MakeSlice(typ Type, len, cap int) Value {
src/reflect/value.go:2931:func MakeChan(typ Type, buffer int) Value {
src/reflect/value.go:2947:func MakeMap(typ Type) Value {
src/reflect/value.go:2953:func MakeMapWithSize(typ Type, n int) Value {
src/reflect/value.go:3438:func makechan(typ *rtype, size int) (ch unsafe.Pointer)
```

Other projects use that idiom it as well:
```
vendor/go.starlark.net/starlark/int.go:43:func MakeInt64(x int64) Int {
vendor/go.starlark.net/starlark/int.go:51:func MakeUint(x uint) Int { return MakeUint64(uint64(x)) }
vendor/go.starlark.net/starlark/int.go:54:func MakeUint64(x uint64) Int {
vendor/go.starlark.net/starlark/int.go:67:func MakeBigInt(x *big.Int) Int {
```

File: pkg/controllers/provisioning/scheduling/topologynodefilter.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/topologynodefilter.go:59]: To me it's about readability of the function when used, the first is far more clear than the second:

```go
// nodes may or may not be considered for counting purposes for topology spread constraints depending on if they
// are selected by the pod's node selectors and required node affinities.  If these are unset, the node always counts.
if !tg.nodeFilter.Matches(node) {
        continue
}
```


```go
// nodes may or may not be considered for counting purposes for topology spread constraints depending on if they
// are selected by the pod's node selectors and required node affinities.  If these are unset, the node always counts.
if err := tg.nodeFilter.Compatible(node); err != nil  {
        continue
}
```

File: pkg/controllers/provisioning/scheduling/topologynodefilter.go
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/topologynodefilter.go:73]: Do we still need this hash function now that it's just an alias type?

Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/topologynodefilter.go]: Good catch, removed

File: pkg/controllers/provisioning/scheduling/topologynodefilter.go
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/topologynodefilter.go:28]: Now that this type has been simplified, it occurs to me that it might be more straightforward to fold this into topologygroup e.g. `nodeFilter []v1alpha5.Requirements`. It should be ~40 lines of code.

**Review by ellistarn — APPROVED** [review]
Let's get one more approver, but I'm good to go.

File: pkg/controllers/provisioning/scheduling/preferences.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/preferences.go:84]: Yes, it can definitely change the result.   There isn't an order specified by K8s, they perform scheduling in a different way from what I can tell (find all compatible nodes, sort by score).  We schedule on the first compatible node treating all preferences as required, and removing one preference at a time until it either schedules or there are no more to remove and it fails.

File: pkg/controllers/provisioning/scheduling/preferences.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/preferences.go:120]: It only gets logged if the pod has failed to schedule.  Most of the time this will never log anything. Preferences is a little odd , look at like 61 or so.  The first time it's called, it doesn't relax anything.

File: pkg/controllers/provisioning/scheduling/queue.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/queue.go:36]: Yes, there are some ideas around "vector bin packing" that I want to investigate at some point that will probably improve upon this.  It will take writing some benchmark code to ensure that we are actually packing more densely, but a different sort order could improve.  

We may also want to sort pods with anti-affinity or topology spreads first.  They are more likely to create new nodes which we can then use to fulfill other scheduling constraints.  Lots of options to investigate and determine what works best.

File: pkg/controllers/provisioning/scheduling/queue.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/queue.go:28]: Scheduling is currently single-threaded and the queue is created to operate on a single batch of pods.  The launching of nodes later on is parallel, but this part isn't. The topology groups, etc. aren't thread safe and making them that way could potentially lead to more scheduling violations.  E.g. multiple goroutines observe the same topology count and then schedule a pod to the same zone, which would now violate a spread.


This problem exists currently as well as we can race kube-scheduler.

File: pkg/controllers/provisioning/scheduling/scheduler.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/scheduler.go:132]: It may vary depending on CPU, but that should be it.  We're just sorting s list of pointers by the pod count which is a constant operation to retrieve.

File: pkg/controllers/provisioning/scheduling/scheduler.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/scheduler.go:90]: > Finding it hard to understand the code, so I'll ask here. Does this code work with an arbitrary amount of zonal topology requirements? Say if we had `5xA, 5xB, 5xC, 5xD, 5xE` pods, would this eventually schedule them with alternating order: `A,B,C,D,E,A,B,C,D,E,A....`?

Yes, but even the example in the comment here is an edge case.  With a normal topology spread, all of the pods in the deployment have the same restrictions.  This case only occurs if you have multiple deployments that use the same topology spread via label selectors, but each deployment has a different zonal restriction.  It's super unlikely and doesn't make sense, but we will still schedule it, though slowly.  The more common case is pod A has affinity to pod B, in which case we need to schedule B and then make one more pass to schedule A.

> Additionally, does this mean the algorithm is roughly `O(n^m)` where `m` is the cardinality of the topologies (e.g. `cardinality({A,B,C,D,E}) = 5`) and `n` is the amount of pod in each?

I don't think it's that bad as we don't care what order we schedule A,B,C,D,E in. We just have to schedule one of {A,B,C,D,E} before we schedule any other.

So if you try to make a bad input case case (using 3 deployments and 3 pods each for simplicity):

AAABBBCCC

First round we schedule A,B,C and then the same in each subsequent round for three rounds of scheduling.  I think any other sort order will schedule faster. I haven't fully looked at it in detail, but gut feel is O(nm) on these weird edge cases.

File: pkg/controllers/provisioning/scheduling/scheduler.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/scheduler.go:111]: Only logs on topology update error, IIRC that error only occurs when a kube-api server call fails which shouldn't occur often.

File: pkg/controllers/provisioning/scheduling/scheduling_benchmark_test.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/scheduling_benchmark_test.go:199]: It's a benchmark test and I left it in for now as I keep adding/removing that line to perform some benchmarking just to get the numbers.

File: pkg/controllers/provisioning/scheduling/topologygroup.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/topologygroup.go:53]: 😂

File: pkg/controllers/provisioning/scheduling/scheduler.go
Comment (ellistarn) [reply pkg/controllers/provisioning/scheduling/scheduler.go:111]: Keep in mind. This call is cached. It should only fail due to a code bug.

File: pkg/controllers/provisioning/scheduling/queue.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/queue.go:67]: Nothing else could use this currently, this is only for bin-packing the pods.  It may be worthwhile moving it to another file if we have multiple strategies and lets users pick, or try multiple and choose the best strategy, but I think it's ok for now.

File: pkg/controllers/provisioning/scheduling/scheduler.go
Comment (njtran) [reply pkg/controllers/provisioning/scheduling/scheduler.go:90]: Thanks for your detailed response. I wanted to make sure that this was going to work for `m>3` topology groups. Seems good to me.

**Comment by BeardyBear** [comment]
Hi, how long would it take for this to be released as part of new Karpenter version?
Is there a way I could run it now or should I just wait for the release?
We really need this feature.

Thank you :)

**Review by ellistarn — APPROVED** [review]

**Comment by tzneal** [comment]
> Hi, how long would it take for this to be released as part of new Karpenter version? Is there a way I could run it now or should I just wait for the release? We really need this feature.
> 
> Thank you :)

We plan to make a snapshot image available soon so it can be tested out before the next release.

**Comment by joebowbeer** [comment]
Also closes #985 ?

**Pull request merged**
